File: src/priv/bajdcc/LALR1/interpret/module/ModuleUI.java
package priv.bajdcc.LALR1.interpret.module;

import priv.bajdcc.LALR1.grammar.Grammar;
import priv.bajdcc.LALR1.grammar.runtime.*;
import priv.bajdcc.LALR1.ui.drawing.UIGraphics;

import java.util.ArrayDeque;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.LinkedBlockingDeque;

/**
 * 【模块】界面
 *
 * @author bajdcc
 */
public class ModuleUI implements IInterpreterModule {

	private static final int INPUT_TIME = 10;
	private static ModuleUI instance = new ModuleUI();
	private UIGraphics graphics;
	private Queue<Character> queue = new LinkedBlockingDeque<>(1024);
	private Queue<Character> queueDisplay = new ArrayDeque<>();
	private StringBuilder sb = new StringBuilder();

	public void setGraphics(UIGraphics graphics) {
		this.graphics = graphics;
	}

	public void addInputChar(char c) {
		queue.add(c);
	}

	public void addDisplayChar(char c) {
		queueDisplay.add(c);
	}

	public static ModuleUI getInstance() {
		return instance;
	}

	@Override
	public String getModuleName() {
		return "sys.ui";
	}

	@Override
	public RuntimeCodePage getCodePage() throws Exception {
		String base = "import \"sys.base\";\n" +
				"import \"sys.list\";\n" +
				"import \"sys.proc\";\n" +
				"import \"sys.string\";\n" +
				"var g_ui_print = func ~(str) {\n" +
				"    var ui_int = call g_create_pipe(\"int#2\");\n" +
				"    foreach (var c : call g_range_string(str)) {\n" +
				"        call g_write_pipe(ui_int, c);\n" +
				"    }\n" +
				"};\n" +
				"export \"g_ui_print\";\n" +
				"var g_ui_printn = func ~(str) {\n" +
				"    call g_ui_print(str);\n" +
				"    call g_ui_println();\n" +
				"};\n" +
				"export \"g_ui_printn\";\n" +
				"var g_ui_println = func ~() {\n" +
				"    call g_ui_print(g_endl);\n" +
				"};\n" +
				"export \"g_ui_println\";\n" +
				"var g_ui_input = func ~() {\n" +
				"    call g_ui_caret(true);\n" +
				"    var h = call g_query_share(\"cmd#histroy\");\n" +
				"    for (;;) {\n" +
				"        var s = call g_ui_input_internal();\n" +
				"        if (!call g_is_null(s)) {\n" +
				"            while (!call g_ui_caret(false)) {};\n" +
				"            call g_ui_println();\n" +
				"            call g_array_add(h, s);\n" +
				"            return s;\n" +
				"        }\n" +
				"        var c = call g_ui_print_input();\n" +
				"        if (!call g_is_null(c)) {\n" +
				"            if (c == '\\ufff0') {\n" +
				"                if (!call g_array_empty(h)) {\n" +
				"                    var old = call g_array_pop(h);\n" +
				"                    call g_ui_input_queue(old);\n" +
				"                }\n" +
				"            } else if (c == '\\uffee') {\n" +
				"                call g_ui_println();\n" +
				"                return call g_ui_input_im();\n" +
				"            } else {\n" +
				"                call g_ui_print(c);\n" +
				"            }\n" +
				"        }\n" +
				"    }\n" +
				"};\n" +
				"export \"g_ui_input\";\n" +
				"var g_ui_inputd = func ~(callback, arr) {\n" +
				"    var handle = call g_wait_pipe(callback);\n" +
				"    while (call g_array_size(arr) == 0) {\n" +
				"        var c = call g_ui_print_input();\n" +
				"        if (!call g_is_null(c)) {\n" +
				"            if (c == '\\uffee') {\n" +
				"                call g_array_add(arr, 'C');\n" +
				"                break;\n" +
				"            }\n" +
				"        }\n" +
				"    }\n" +
				"    call g_write_pipe(handle, call g_array_get(arr, 0));\n" +
				"};\n" +
				"export \"g_ui_inputd\";\n";

		Grammar grammar = new Grammar(base);
		RuntimeCodePage page = grammar.getCodePage();
		IRuntimeDebugInfo info = page.getInfo();
		buildUIMethods(info);

		return page;
	}

	private void buildUIMethods(IRuntimeDebugInfo info) {
		info.addExternalFunc("g_ui_print_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "显示输出";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kChar};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				graphics.drawText((char) args.get(0).getObj());
				return null;
			}
		});
		info.addExternalFunc("g_ui_input_internal", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "显示输入";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				status.getService().getProcessService().sleep(status.getPid(), INPUT_TIME);
				Character c = queue.poll();
				if (c == null) {
					return null;
				}
				if (c.equals('\n')) {
					String str = sb.toString();
					sb = new StringBuilder();
					queueDisplay.clear();
					return new RuntimeObject(str);
				} else {
					if (c < '\ufff0') {
						sb.append(c);
					}
					queueDisplay.add(c);
				}
				return null;
			}
		});
		info.addExternalFunc("g_ui_input_im", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "立即显示输入";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				status.getService().getProcessService().sleep(status.getPid(), INPUT_TIME);
				String str = sb.toString();
				sb = new StringBuilder();
				queueDisplay.clear();
				return new RuntimeObject(str);
			}
		});
		info.addExternalFunc("g_ui_input_queue", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "显示输入缓冲";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				status.getService().getProcessService().sleep(status.getPid(), INPUT_TIME);
				String str = String.valueOf(args.get(0).getObj());
				sb.append(str);
				for (char c : str.toCharArray()) {
					queueDisplay.add(c);
				}
				return null;
			}
		});
		info.addExternalFunc("g_ui_print_input", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "实时显示输入";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				return new RuntimeObject(queueDisplay.poll());
			}
		});
		info.addExternalFunc("g_ui_caret", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "设置光标闪烁";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kBool};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				boolean caret = (boolean) args.get(0).getObj();
				if (caret) {
					graphics.setCaret(caret);
					return null;
				} else {
					graphics.setCaret(caret);
					status.getService().getProcessService().sleep(status.getPid(), INPUT_TIME);
					return new RuntimeObject(graphics.isHideCaret());
				}
			}
		});
	}
}

File: src/main/java/com/bajdcc/LALR1/interpret/module/ModuleUI.kt
package com.bajdcc.LALR1.interpret.module

import com.bajdcc.LALR1.grammar.Grammar
import com.bajdcc.LALR1.grammar.runtime.*
import com.bajdcc.LALR1.ui.drawing.UIGraphics
import com.bajdcc.util.ResourceLoader
import java.util.*
import java.util.concurrent.LinkedBlockingDeque
import javax.swing.JPanel

/**
 * 【模块】界面
 *
 * @author bajdcc
 */
class ModuleUI : IInterpreterModule {
    private var graphics: UIGraphics? = null
    private var panel: JPanel? = null
    private val queue = LinkedBlockingDeque<Char>(10000)
    private val queueDisplay = ArrayDeque<Char>()
    private var sb = StringBuilder()
    private var runtimeCodePage: RuntimeCodePage? = null

    override val moduleName: String
        get() = "sys.ui"

    override val moduleCode: String
        get() = ResourceLoader.load(javaClass)

    override val codePage: RuntimeCodePage
        @Throws(Exception::class)
        get() {
            if (runtimeCodePage != null)
                return runtimeCodePage!!

            val base = ResourceLoader.load(javaClass)

            val grammar = Grammar(base)
            val page = grammar.codePage
            val info = page.info
            buildUIMethods(info)

            runtimeCodePage = page
            return page
        }

    fun setGraphics(graphics: UIGraphics) {
        this.graphics = graphics
    }

    fun setPanel(panel: JPanel) {
        this.panel = panel
    }

    fun addInputChar(c: Char) {
        queue.add(c)
    }

    fun addDisplayChar(c: Char) {
        queueDisplay.add(c)
    }

    private fun buildUIMethods(info: IRuntimeDebugInfo) {
        info.addExternalFunc("g_ui_print_internal_block",
                RuntimeDebugExec("显示输出", arrayOf(RuntimeObjectType.kChar))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val failure = graphics!!.drawText(args[0].char)
                        if (failure) {
                            status.service.processService.waitForUI()
                            status.service.processService.sleep(status.pid, PRINT_BLOCK_TIME)
                        }
                        RuntimeObject(failure)
                    }
                })
        info.addExternalFunc("g_ui_input_internal",
                RuntimeDebugExec("显示输入")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        status.service.processService.sleep(status.pid, INPUT_TIME)
                        val c = queue.poll()
                        when (c) {
                            null -> null
                            '\n' -> {
                                val str = sb.toString()
                                sb = StringBuilder()
                                queueDisplay.clear()
                                RuntimeObject(str)
                            }
                            '\b' -> {
                                if (sb.isNotEmpty())
                                    sb.deleteCharAt(sb.length - 1)
                                graphics!!.drawText('\b')
                                null
                            }
                            else -> {
                                if (c < '\ufff0') {
                                    sb.append(c)
                                }
                                queueDisplay.add(c)
                                null
                            }
                        }
                    }
                })
        info.addExternalFunc("g_ui_input_im",
                RuntimeDebugExec("立即显示输入")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        status.service.processService.sleep(status.pid, INPUT_TIME)
                        val str = sb.toString()
                        sb = StringBuilder()
                        queueDisplay.clear()
                        RuntimeObject(str)
                    }
                })
        info.addExternalFunc("g_ui_input_queue",
                RuntimeDebugExec("显示输入缓冲", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        status.service.processService.sleep(status.pid, INPUT_TIME)
                        val str = args[0].obj.toString()
                        sb.append(str)
                        for (c in str.toCharArray()) {
                            queueDisplay.add(c)
                        }
                        null
                    }
                })
        info.addExternalFunc("g_ui_print_input",
                RuntimeDebugExec("实时显示输入")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        status.service.processService.sleep(status.pid, INPUT_TIME)
                        RuntimeObject(queueDisplay.poll())
                    }
                })
        info.addExternalFunc("g_ui_caret",
                RuntimeDebugExec("设置光标闪烁", arrayOf(RuntimeObjectType.kBool))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val caret = args[0].bool
                        if (caret) {
                            graphics!!.setCaret(true)
                            null
                        } else {
                            graphics!!.setCaret(false)
                            status.service.processService.sleep(status.pid, INPUT_TIME)
                            RuntimeObject(graphics!!.isHideCaret)
                        }
                    }
                })
        info.addExternalFunc("g_ui_fallback",
                RuntimeDebugExec("撤销上次输入")
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        sb = StringBuilder()
                        graphics!!.fallback()
                        null
                    }
                })
        info.addExternalFunc("g_ui_cols",
                RuntimeDebugExec("多少列")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(graphics!!.cols.toLong()) })
        info.addExternalFunc("g_ui_rows",
                RuntimeDebugExec("多少行")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(graphics!!.rows.toLong()) })
        info.addExternalFunc("g_ui_text_length",
                RuntimeDebugExec("一段文字宽度", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val str = args[0].obj.toString()
                        RuntimeObject(graphics!!.calcWidth(str).toLong())
                    }
                })
        info.addExternalFunc("g_ui_create_dialog_internal",
                RuntimeDebugExec("创建对话框", arrayOf(RuntimeObjectType.kString, RuntimeObjectType.kString, RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val caption = args[0].obj.toString()
                        val text = args[1].obj.toString()
                        val mode = args[2].long
                        RuntimeObject(status.service.dialogService.create(caption, text, mode.toInt(), panel!!))
                    }
                })
        info.addExternalFunc("g_ui_show_dialog_internal",
                RuntimeDebugExec("弹出对话框", arrayOf(RuntimeObjectType.kPtr))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        RuntimeObject(status.service.dialogService.show(handle))
                    }
                })
    }

    companion object {

        private const val INPUT_TIME = 10
        val instance = ModuleUI()
        private const val PRINT_BLOCK_TIME = 5
    }
}
